R/utils.R

```diff
@@ -17,24 +17,47 @@ formatSolveBioResponse <- function (res, raw = FALSE) {
         res$'_url' = url
         return(res)
     }
+
 }
 
-formatSolveBioQueryResponse <- function (res, raw = FALSE, row.names = NULL) {
+formatSolveBioQueryResponse <- function (id, res, raw = FALSE, row.names = NULL) {
     # res will be the output of formatSolveBioResponse
     if (!raw & is.data.frame(res$results)) {
-        # Flatten the data frame
+        # Append column "_id" if it's not already there
+        if (!"_id" %in% row.names) {
+            row.names <- c(row.names, "_id")
+        }
+        # Remove columns from row.names that are not in the results query
+        diff <- setdiff(row.names, colnames(res$results))
+        row.names <- row.names [! row.names %in% diff]
+
+        # Flatten the data frame and sort columns by row.names
         res$results <- jsonlite::flatten(res$results)
+        res$results <- res$results[, row.names]
+
+        # Create a dataframe that maps column names with titles
+        names <-  do.call(Dataset.fields, list(id, limit=200))$data$name # get fields name
```

Comment:
This function is shared by Object query and Dataset query. Two requests:

1. Can these be done in the Dataset.data function?
2. Can these requests be combined in to one?

Reply:
Also if row.names is passed in, why do we need to do this again?

Ideally we just pass in (optionally) the list of fields (names and titles) in the correct order.

Reply:
1. Yes. Now all formating of the results for the `Dataset.query` is done inside `Dataset.data` (This formating was actually not needed for the formating of the Object). Since all formatting is done inside `Dataset.data`, I'm just flattening the result in the `formatSolveBioQueryResponse` function
2. `col.names` are now only used in the case of `Object.data`, so I've basically just ensured that it enters the correct `if` branch

Instead of passing `col.names` as a list of names only, I've changed it a bit, to send `Dataset.fields(id, limit=200)$data` and then I extract list of names and titles from there. What do you think?